### PR TITLE
[connectors] feature: Upsert table parents from managed ds

### PR DESCRIPTION
**Pull request by tdraier**
## Description

- Update connectors to upsert table parents for googledrive / notions / microsoft
- Add cli to get item parents
- Add migration script
- Core api to update parents

## Deploy Plan

deploy `core`, `front`, then `connectors`
run migration script

### CODE REVIEW

File: core/src/stores/postgres.rs [hunk 1]
```diff
@@ -2273,7 +2273,7 @@ impl Store for PostgresStore {
                     timestamp, tags_array, parents) \
                    VALUES (DEFAULT, $1, $2, $3, $4, $5, $6, $7, $8) \
                    ON CONFLICT (table_id, data_source) DO UPDATE \
-                   SET name = EXCLUDED.name, description = EXCLUDED.description \
+                   SET name = EXCLUDED.name, description = EXCLUDED.description, tags_array = EXCLUDED.tags_array, parents = EXCLUDED.parents \
```
Comment (spolu) [inline core/src/stores/postgres.rs:2276]: Good catch. Missing timestamp no?

Comment (tdraier) [reply core/src/stores/postgres.rs]: yes :-)

**Review by flvndvd — APPROVED** [review]

File: front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import { getDataSource } from "@app/lib/api/data_sources";
+import { Authenticator, getAPIKey } from "@app/lib/auth";
+import logger from "@app/logger/logger";
+import { apiError, withLogging } from "@app/logger/withlogging";
+
+export type PostParentsResponseBody = {
+  updated: true;
+};
+
+/**
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts:15]: This endpoint should not be documented as we don't want to open the table related endpoints yet.

File: front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "POST":
+      if (
+        !req.body ||
+        !Array.isArray(req.body.parents) ||
+        !req.body.parents.every((p: any) => typeof p === "string")
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts:112]: Can we use io-ts here?

File: connectors/src/connectors/notion/index.ts [hunk 4]
```diff
@@ -601,12 +601,16 @@ export class NotionConnectorManager extends BaseConnectorManager<null> {
...
+      ? internalId.substring("notion-".length)
+      : internalId;
+
     const memo = memoizationKey || uuidv4();
 
     try {
       const parents = await getParents(
         this.connectorId,
-        internalId,
+        id,
```
Comment (flvndvd) [inline connectors/src/connectors/notion/index.ts:613]: Isn't this a breaking change with what we have today? This `getParents` function relies on a Redis cache, what ids do we store there?

File: connectors/src/connectors/microsoft/temporal/spreadsheets.ts [hunk 5]
```diff
@@ -165,11 +169,19 @@ async function processSheet(
   if (rawHeaders && rows.length > 1) {
     const headers = getSanitizedHeaders(rawHeaders);
 
+    const [, ...parents] = await getParents({
```
Comment (flvndvd) [inline connectors/src/connectors/microsoft/temporal/spreadsheets.ts:172]: Same why skipping the first element?

File: connectors/src/connectors/microsoft/lib/cli.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import type {
   AdminSuccessResponseType,
-  MicrosoftCheckFileResponseType,
+  CustomDataResponseType,
```
Comment (flvndvd) [inline connectors/src/connectors/microsoft/lib/cli.ts:3]: IMHO, this type's name isn't explicit enough.
```suggestion
  CheckFileGenericResponseType,
```

File: connectors/src/api/get_connector_permissions.ts [hunk 7]
```diff
@@ -82,6 +82,10 @@ const _getConnectorPermissions = async (
     viewType,
   });
 
+  if (pRes.isOk()) {
+    console.log("================nodes", pRes.value);
+  }
+
```
Comment (flvndvd) [inline connectors/src/api/get_connector_permissions.ts:88]: ✂️

File: connectors/src/connectors/notion/index.ts (hunk 4 above)
Comment (tdraier) [reply connectors/src/connectors/notion/index.ts:613]: No, the only difference is that we support id starting with "notion-" (used only in tables) - if the id is a "standard" notion id it's the same value

File: front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts (hunk 2 above)
Comment (tdraier) [reply front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts:15]: eslint says all endpoints need to be documented ? (and all here indeed)

File: connectors/src/connectors/microsoft/temporal/spreadsheets.ts (hunk 5 above)
Comment (tdraier) [reply connectors/src/connectors/microsoft/temporal/spreadsheets.ts:172]: `getParents` return itself as the first element (method name is confusing indeed, and we have the same for google and notion) - probably useful for `connector.retrieveContentNodeParents`, but here we only need the parents list

File: front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts (hunk 2 above)
Comment (flvndvd) [reply front/pages/api/v1/w/[wId]/data_sources/[name]/tables/[tId]/parents.ts:15]: Except for the one around tables 🙃.

File: connectors/src/connectors/notion/index.ts (hunk 4 above)
Comment (flvndvd) [reply connectors/src/connectors/notion/index.ts:613]: So databases (tables) in Notion have an id that starts with `notion-` while for documents no? If that's the case then IMHO it's worth a comment.

File: connectors/src/connectors/microsoft/temporal/spreadsheets.ts (hunk 5 above)
Comment (flvndvd) [reply connectors/src/connectors/microsoft/temporal/spreadsheets.ts:172]: Can we either create a function for it or add a parameter like `excludeSelf`?

File: connectors/src/connectors/notion/index.ts (hunk 4 above)
Comment (tdraier) [reply connectors/src/connectors/notion/index.ts]: > So databases (tables) in Notion have an id that starts with `notion-` while for documents no? If that's the case then IMHO it's worth a comment.

we can forget about it, it was only a hack for cli usage with notion-* ids, but does make sense anyway

File: connectors/migrations/20240802_table_parents.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+  getTable,
+  updateTableParentsField,
+} from "@connectors/lib/data_sources";
+import { GoogleDriveSheet } from "@connectors/lib/models/google_drive";
+import { MicrosoftNodeModel } from "@connectors/lib/models/microsoft";
+import { NotionDatabase } from "@connectors/lib/models/notion";
+import type { Logger } from "@connectors/logger/logger";
+import { ConnectorResource } from "@connectors/resources/connector_resource";
+
+// Deleting all existing Google Drive CSV files
```
Comment (flvndvd) [inline connectors/migrations/20240802_table_parents.ts:20]: ✂️

File: connectors/migrations/20240802_table_parents.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    if (execute) {
+      await updateTableParentsField({ tableId, parents, dataSourceConfig });
+    }
+  }
+}
+
+export async function microsoftTables(
+  connector: ConnectorResource,
+  check: boolean,
+  execute: boolean,
```
Comment (flvndvd) [inline connectors/migrations/20240802_table_parents.ts:62]: `nit` You could have relied only on `execute`.

File: connectors/src/lib/data_sources.ts [hunk 10]
```diff
@@ -747,6 +793,41 @@ export async function deleteTableRow({
...
+  let dustRequestResult: AxiosResponse;
+  try {
+    dustRequestResult = await axiosWithTimeout.get(endpoint, dustRequestConfig);
+  } catch (e) {
+    localLogger.error({ error: e }, "Error getting structured data from Dust.");
+    throw e;
+  }
+
+  if (dustRequestResult.status === 404) {
+    localLogger.info("Structured data doesn't exist on Dust. Ignoring.");
```
Comment (flvndvd) [inline connectors/src/lib/data_sources.ts:824]: We go with tables and not structured data 🙃.

**Review by flvndvd — APPROVED** [review]
LGTM, assuming you've tested 🙏

**Pull request merged**